### PR TITLE
Google Analytics 4 For WPcom 

### PR DESCRIPTION
**Pull request by gmovr**
#### Proposed Changes

* Implementing Google Analytics 4 for WPcom
* Will handle page views, generic events, and purchase tracking.
* Related to: pau2Xa-48B-p2 
* Basically, the code is made more generic from the previous implementation for Jetpack Cloud. This allows us to re-use the functionality for WPcom (and potentially other sites down the line).

#### Notes
* Purchases for Jetpack and WPcom are fired on the WPcom implementation. This is in line with the current implementation of Enhanced Ecommerce tracking in Universal Analytics. I've updated the PR to reflect this, and making the product brand independent of the GA4 tracking environment. This also takes into account some "edge cases" e.g. if the cart has both a WPcom and Jetpack product, both will be parsed. One example would be https://wordpress.com/checkout/jetpack_search,premium/site_slug. Which will generate the cart items below:

```javascript
(2) [{…}, {…}]
0: {product_id: 1003, billing_plan_id: '16615', product_name: 'WordPress.com Premium', product_name_en: 'WordPress.com Premium', product_slug: 'value_bundle', …}
1: {product_id: 800, billing_plan_id: '20459', product_name: 'Jetpack Search', product_name_en: 'Jetpack Search', product_slug: 'wpcom_search', …}
length: 2
[[Prototype]]: Array(0)
```

#### Testing Instructions
It may be helpful to open the network tab in developer tools, and from there filter on `G-1H4VG5F5JF` (which is the measurement ID we're using for GA4).

* Sandbox store (purchase events won't fire when using free credits, as cost < $0.01).
* Allow ad tracking using config flag `?flags=gdpr-banner,google-analytics,ad-tracking`.
    * If you're in EU, you can accept the cookie banner -- or set the `sensitive_pixel_option` cookie manually. 
* Check that gtag.js is initializing correctly. You should see a call with the parameter `config`.
* Check that page views are firing (event name `page_view`).
* Buy a plan, verify that purchase is triggering (event name `purchase`).
* Automated tests pass.
* No TS/console errors.
* If you have access to the GA4 property, you can also verify that the purchase is showing under real-time. 
* **Verify that the tracking works on Jetpack Cloud, by going through the same test procedure as above.**

#### Pre-merge Checklist

Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in [Simple](P9HQHe-k8-p2-p2), [Atomic](P9HQHe-jW-p2-p2), and [self-hosted Jetpack sites](PCYsg-g6b-p2-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Are we memoizing when appropriate (for expensive computations)? More info in [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md) and [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors)
- [ ] Have we sent any new strings [for translation](PCYsg-1vr-p2-p2) ASAP?

Resolves Automattic/martech#827 Automattic/martech#831 Automattic/martech#832

### CODE REVIEW

File: client/lib/analytics/ad-tracking/google-analytics-4.ts
Comment (tyxla) [inline client/lib/analytics/ad-tracking/google-analytics-4.ts]: Having `shouldSendToJetpack` as a boolean flag feels like a code smell and a limitation. Right now we have 2 entities to record GA for - Jetpack and WP.com, but what will happen if we have 3? With that in mind, isn't it better to have this argument accept a string (`wpcom` or `jetpack`)? That would allow adding more entities to track analytics in the future, if necessary.

**Review by tyxla — COMMENTED** [review]
From a frameworks / code perspective, this looks good overall 👍 

My only reservation is regarding keeping the boolean flag to decide whether to record to Jetpack or WP.com's analytics, which seems to lack flexibility.

For testing and final approval, I'll defer to the team working on this.

File: client/lib/analytics/ad-tracking/google-analytics-4.ts
Comment (niranjan-uma-shankar) [inline client/lib/analytics/ad-tracking/google-analytics-4.ts]: Should we also change this function name to a generic one like `fireEcommerceAddToCart` , since it applies to both WPCOM and Jetpack? 

Also, I couldn't find `fireJetpackEcommerceAddToCart()` being used anywhere in the code, could you please check its usage?

**Review by niranjan-uma-shankar — COMMENTED** [review]
Could you help me understand if we also need to do GA4 tracking for other event types, which are handled in `fireGoogleAnalyticsEvent()`?

https://github.com/Automattic/wp-calypso/blob/892ae1abb1133596d499b1d573c21c0eb198c0ba/client/lib/analytics/ad-tracking/google-analytics.js#L101-L106

For example, should an event like `calypso_signup_complete` which is tracked in GA [here](https://github.com/Automattic/wp-calypso/blob/7b9c1326c3210ce18813f2296f4599830510f5dd/client/lib/analytics/signup.js#L87) also get tracked in GA4?

File: client/lib/analytics/ad-tracking/record-order.js
Comment (niranjan-uma-shankar) [inline client/lib/analytics/ad-tracking/record-order.js]: This function is named `fireEcommercePurchaseGA4()` - is it intended to track only ecommerce plan purchases? However, the function in which it is housed implies that we track all purchases/orders in WPCOM. So which one of these is applicable?

File: client/lib/analytics/ad-tracking/record-order.js
Comment (niranjan-uma-shankar) [inline client/lib/analytics/ad-tracking/record-order.js:84]: Since the function below is called `recordOrderInWPcomGA4`, for consistency sake should we also rename this to `recordOrderInJetpackGA4 `?

File: client/lib/analytics/ad-tracking/record-order.js
Comment (niranjan-uma-shankar) [inline client/lib/analytics/ad-tracking/record-order.js:509]: My preference would be to combine `recordOrderInJetpackGA()` and `recordOrderInWPcomGA4()` into a single function, since both functions have almost the same lines of code. The following line would be the only change needed:

```
const sendToJetpack = wpcomJetpackCartInfo.containsJetpackProducts ? true : false;

```

File: client/lib/analytics/ad-tracking/google-analytics-4.ts
Comment (gmovr) [reply client/lib/analytics/ad-tracking/google-analytics-4.ts]: Thanks for the review!

For the generic events (that don't have an explicit `send_to` property in the object), they will send to every gtag which is in the dataLayer. I.e. if you check the `window.dataLayer` object, we init gtag for DC, Google Ads, Google Analytics, and more. :) If we don't have the `send_to` set, it will pass this event to all the gtag that are configured, while the send_to will send to a specific conversion ID (for a specific account).

Since both gtag (which is currently used for i.e. Google Ads), is also used for the GA4 integration, those will be picked up. I.e. you can see them being sent to the gtag ID `G-1H4VG5F5JF` (for user_interaction, signup start, etc.).

The events such as `fireJetpackEcommerceAddToCart()` aren't referenced anywhere. I saw them existing in the file, and they'll be referenced in a follow-up for tracking the "add to cart" event with a cart, so we can see i.e. how many users add a product to the cart, and end up buying, to see drop-offs. 

Renaming that one is a good idea, I'll do that! 👍 

We do track the "add to cart" event (via `fireGoogleAnalyticsEvent`), but it's just a "generic" event -- and we'd need a more specific one parsing the cart in a follow-up (more similar to the purchase event).

File: client/lib/analytics/ad-tracking/record-order.js
Comment (gmovr) [reply client/lib/analytics/ad-tracking/record-order.js]: Actually, reading it like that -- the naming is a bit confusing! It's because this event takes in the cart, and allows us to track with the Ecommerce views in Google Analytics. Those take in a cart, price, etc. So it's not just for ecommerce sales, but for any sales (that allow us to use the ecommerce views/reports).

File: client/lib/analytics/ad-tracking/record-order.js
Comment (gmovr) [reply client/lib/analytics/ad-tracking/record-order.js:84]: I agree! :)

File: client/lib/analytics/ad-tracking/record-order.js
Comment (gmovr) [reply client/lib/analytics/ad-tracking/record-order.js:509]: This one is actually an old one: `recordOrderInJetpackGA` and also handles universal analytics. For WPcom, ecommerce has been tracked in Universal Analytics, but using the Enhanced Ecomerce function/reporting.
https://github.com/Automattic/wp-calypso/blob/892ae1abb1133596d499b1d573c21c0eb198c0ba/client/lib/analytics/ad-tracking/record-order.js#L439

So I made a new function for GA4 for WPcom, to separate the concerns.

I'll have a think about this one, and see if we can approach it differently.

File: client/lib/analytics/ad-tracking/google-analytics-4.ts
Comment (robertsreberski) [reply client/lib/analytics/ad-tracking/google-analytics-4.ts]: I second that; it sounds like a good idea. Especially since the code is written in Typescript and we can explicitly define enums.

File: client/lib/analytics/utils/product-to-ga-item.ts
Comment (robertsreberski) [inline client/lib/analytics/utils/product-to-ga-item.ts]: Cannot we distinguish the brand somehow from the `ResponseCartProduct` rather than passing the artificial flag `sendToJetpack`?

File: client/lib/analytics/utils/cart-to-ga-purchase.ts
Comment (robertsreberski) [inline client/lib/analytics/utils/cart-to-ga-purchase.ts]: I'd rather create a standalone `cartInfo` object type here (that would have just `costUSD` property, and `cartItems`), rather than pass an additional `sendToJetpack` flag

File: client/lib/analytics/ad-tracking/google-analytics-4.ts
Comment (robertsreberski) [inline client/lib/analytics/ad-tracking/google-analytics-4.ts]: Just leaving here the remainder that we should make wpcom marketing team aware of the change in `page_view` events. Currently, wpcom property collets both cloud.jetpack.com and wordpress.com views.

Reference: https://github.com/Automattic/wp-calypso/blob/0e3346c8c6c3d0e59664110ff790da243e9b4bda/client/lib/analytics/ad-tracking/google-analytics.js#L86

File: client/lib/analytics/utils/product-to-ga-item.ts
Comment (gmovr) [reply client/lib/analytics/utils/product-to-ga-item.ts]: I agree, passing this artificial property down through the functions isn't the cleanest way.

File: client/lib/analytics/ad-tracking/google-analytics-4.ts
Comment (gmovr) [reply client/lib/analytics/ad-tracking/google-analytics-4.ts]: Thanks, this is good feedback. The Universal Analytics property is a bit "misleading" I think. I.e. we split out conversions in the code, but not page views. 

I also think the current UA code isn't 100% correct; it's not fully functional (I don't see hostnames containing jetpack in the UA view), all though we do send them. 

I think it's because of the cross-domain setup, that we ran earlier when we had discover.wordpress.com, blog, etc. Those are all in sub-directories now (example: wordpress.com/discover), while earlier they had their own views.

I might be wrong though -- but @debbiengWP and I will verify the data once this is deployed.

File: client/lib/analytics/ad-tracking/google-analytics.js
Comment (robertsreberski) [inline client/lib/analytics/ad-tracking/google-analytics.js:78]:
```suggestion
	const trackingEnvironment = useJetpackGoogleAnalytics
```

How about changing the name of this variable? I think it's more accurate.

File: client/lib/analytics/utils/product-to-ga-item.ts
Comment (gmovr) [reply client/lib/analytics/utils/product-to-ga-item.ts]: Thanks! I've made the brand here independent of the tracking environment. I realized that these shouldn't be coupled together, as we'er currently also firing Jetpack product orders on our Enhanced Ecommerce tracking (Universal Analytics). 

We're now parsing the brand from the cart items. And the tracking environment is only for one thing (which Gtag to send the event to), with no other side effects on the logic.

File: client/lib/analytics/ad-tracking/google-analytics.js
Comment (gmovr) [reply client/lib/analytics/ad-tracking/google-analytics.js:78]: I've refactored these to `ga4PropertyGtag` so they're all the same. I think it's more descriptive, as it's a tracking environment specifically for the Gtag implementations of GA4. :)

File: client/lib/analytics/utils/cart-to-ga-purchase.ts
Comment (gmovr) [inline client/lib/analytics/utils/cart-to-ga-purchase.ts]: @robertsreberski -- I think we can remove this conditional now, since we're already checking for `wpcomJetpackCartInfo.containsJetpackProducts` in `recordOrderInJetpackGA` in `record-order.js`. Will simplify it.

I'm wondering if there are instances when the cart would include both a wpcom plan and a jetpack product/plan, I'm sure there are, and then this assumption may be wrong. I.e. users buying Jetpack products on WPcom, or WPcom products on Jetpack cloud 🤔 

If not removing it, making the conditonal only for Jetpack, and allowing all on WPcom for correct product item and cost parsing.

Comment (debbiengWP) [reply client/lib/analytics/utils/cart-to-ga-purchase.ts]: Hi @gmovr! Just saw that you are updating cart items. Can you please include product type? e.g. product type: wp-bundle, domain, difm. Plan type (personal, premium), frequency (monthly, annual) and amount associated? Also the amount is net (includes discount right). Coupon codes will be awesome too.

File: client/lib/analytics/ad-tracking/record-order.js
Comment (robertsreberski) [inline client/lib/analytics/ad-tracking/record-order.js]:
```suggestion
			Ga4PropertyGtag.JETPACK
```

I think it might be a mistake

File: client/lib/analytics/utils/cart-to-ga-purchase.ts
Comment (robertsreberski) [reply client/lib/analytics/utils/cart-to-ga-purchase.ts]: > If not removing it, making the conditonal only for Jetpack, and allowing all on WPcom for correct product item and cost parsing.

I think this option sounds the best. I would still keep this conditional in jetpack env for the bulletproof solution.

Comment (gmovr) [reply client/lib/analytics/utils/cart-to-ga-purchase.ts]: @debbiengWP - Yes, I think we should do this in a follow-up. I need to research more how we can use this meta data -- but I definitely think this is a good idea.

File: client/lib/analytics/ad-tracking/record-order.js
Comment (gmovr) [reply client/lib/analytics/ad-tracking/record-order.js]: Yes, definitely. Updating. :)

**Comment by gmovr** [comment]
^ Rebase for tests

File: client/lib/analytics/ad-tracking/google-analytics-4.ts
Comment (robertsreberski) [inline client/lib/analytics/ad-tracking/google-analytics-4.ts]: Shouldn't it be:

```suggestion
		send_to: ga4Properties[ ga4PropertyGtag ],
```

File: client/lib/analytics/test/google-analytics-4.js
Comment (robertsreberski) [inline client/lib/analytics/test/google-analytics-4.js]: Unused comment?

File: client/lib/analytics/test/google-analytics-4.js
Comment (robertsreberski) [inline client/lib/analytics/test/google-analytics-4.js:38]: I understood the sense of this test as one checking if we do NOT fire Jetpack Gtag rather than just checking that WPcom has been fired.

I would add the assertion:
```
expect( window.gtag ).not.toHaveBeenCalledWith(
 'config',
 TRACKING_IDS.jetpackGoogleGA4Gtag,
 params
);
```

What do you think?

**Comment by robertsreberski** [comment]
Looking almost great, a few nitpicks from my side to ensure no misunderstandings or bugs.

File: client/lib/analytics/ad-tracking/google-analytics-4.ts
Comment (gmovr) [reply client/lib/analytics/ad-tracking/google-analytics-4.ts]: You're right! Thanks!

File: client/lib/analytics/test/google-analytics-4.js
Comment (gmovr) [reply client/lib/analytics/test/google-analytics-4.js]: I was testing that approach first, but realized the new approach was better. Removing it now.

File: client/lib/analytics/test/google-analytics-4.js
Comment (gmovr) [reply client/lib/analytics/test/google-analytics-4.js:38]: That's a good idea. I've added that assertion as well now. I will work more on test coverage of the lib, since we'll do more refactors -- and we can do those more confidently with more tests. :)

**Comment by gmovr** [comment]
Thanks for the detailed reviews @robertsreberski -- I've pushed a few more commits to address the input above 👍

File: client/lib/analytics/test/google-analytics-4.js
Comment (robertsreberski) [inline client/lib/analytics/test/google-analytics-4.js:54]: One small idea, when during testing we care only about a specific param being matched, here are the first two ones, we can ignore the rest so that our assertion will not "succeed falsely"
```suggestion
			expect( window.gtag ).not.toHaveBeenCalledWith(
				'config',
				TRACKING_IDS.jetpackGoogleGA4Gtag,
				expect.anything()
			);
```

**Review by robertsreberski — COMMENTED** [review]
LGTM, one small idea for the unit tests improvement.

I'll give it a final test on my local environment today and then approve the PR if everything succeedes.

**Review by robertsreberski — APPROVED** [review]
Tested the implementation locally using Google Tag Assistant (https://tagassistant.google.com/) and all the WPcom implementations (on both calypso and jetpack-cloud environments) work as expected!

With one small remark regarding the Jetpack Pricing page (https://cloud.jetpack.com/pricing) that appears to send page_view events to the WPcom environment. However, it looks like a legacy issue not connected to the changes introduced in this PR, so I don't see it as a blocker.

**Pull request merged**
